exec/cnex/nstring.c

```diff
@@ -208,6 +222,111 @@ size_t string_findChar(TString *self, char c)
     return NPOS;
 }
 
+int64_t string_findFirstOf(TString *self, size_t pos, const char *str)
+{
+    if (self->data == NULL || self->length == 0 || pos > self->length || str == NULL) {
+        return -1;
+    }
+
+    for (size_t l = pos; l < self->length; l++) {
+        if (strchr(str, self->data[l]) != NULL) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+static inline BOOL isNotOneOf(TString *self, char c)
+{
+    for (size_t i = 0; i < self->length; i++) {
+        if (self->data[i] == c) {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+int64_t string_findFirstNotOf(TString *self, TString *p)
+{
+    if (self->data == NULL || self->length == 0 || p == NULL || p->data == NULL) {
+        return -1;
+    }
+
+    // If a string of no chars is passed, should we return 0, or -1?  Technically, we didn't find anything,
+    // so the first char we didn't find would be 0.
+    if (p->length == 0) {
+        return 0;
+    }
+
+    for (size_t l = 0; l < self->length; l++) {
+        if (isNotOneOf(p, self->data[l])) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+int64_t string_findLastNotOf(TString *self, TString *p)
+{
+    if (self->data == NULL || self->length == 0) {
+        return -1;
+    }
+
+    if (p == NULL || p->data == NULL || p->length == 0) {
+        return self->length;
+    }
+
+    for (size_t l = self->length; l > 0; l--) {
+        if (isNotOneOf(p, self->data[l-1])) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+int64_t string_findString(TString *self, size_t pos, TString *p)
+{
+    int64_t r = -1;
+    if (p == NULL || p->data == NULL || p->length == 0) {
+        return r;
+    }
+
+    if (self == NULL || self->data == NULL || self->length == 0 || pos > self->length) {
+        return r;
+    }
+
+    for (size_t l = pos; l < self->length; l++) {
+        if (self->data[l] == p->data[0]) {
+            r = l;
+            if (self->length < (l + p->length)) {
+                // The match cannot possibly exist if we don't have enough string left for the pattern to be in.
+                return -1;
+            }
+            BOOL bFound = TRUE;
+            for (size_t n = 0; n < p->length; n++) {
+                if (self->data[l+n] != p->data[n]) {
+                    bFound = FALSE;
+                    break;
+                }
+            }
+            if (bFound) {
+                return r;
+            }
+        }
+    }
+    return -1;
+}
+
+int64_t string_findCString(TString *self, const char *p)
+{
+    int64_t r = -1;
+
+    TString *psz = string_createCString(p);
+    r = string_findString(self, 0, psz);
+    string_freeString(psz);
+    return r;
+}
```

Comment:
You don't need this, the only place you're calling it is from unit test code. Remove it.

Reply:
I'd like to leave it in here.  I was using it at some point with some other code, but later took it out.  This piece of code could come in handy later on, not to mention in the unit test code.  I guess at the very least I can throw this in the unit test code file, and just use it from there.

Reply:
Moved this chunk of code to test_string_support.c